File: MishkatulIlm-Server/Services/LessonFrequencyRules.cs
namespace MishkatulIlm_Server.Services;

public static class LessonFrequencyRules
{
    private static readonly string[] Ranked =
    [
        "BIWEEKLY",
        "ONCE-WEEK",
        "TWICE-WEEK",
        "THREE-WEEK",
        "FOUR-PLUS-WEEK",
    ];

    public static bool TryValidate(string lessonFrequency, int subjectCount, out string? error)
    {
        error = null;
        var code = lessonFrequency.Trim().ToUpperInvariant();
        if (string.IsNullOrEmpty(code))
        {
            error = "Lesson frequency is required.";
            return false;
        }

        if (subjectCount != 3)
            return IsKnownFrequency(code);

        if (code == "FLEXIBLE")
        {
            error = "Flexible frequency is not available when three subjects are selected.";
            return false;
        }

        if (code is "TWICE-WEEK" or "THREE-WEEK")
            return true;

        error = "With 3 subjects, choose twice or three times per week.";
        return false;
    }

    private static bool IsKnownFrequency(string code) =>
        RankOf(code) is not null || code == "FLEXIBLE";

    private static int? RankOf(string code)
    {
        for (var i = 0; i < Ranked.Length; i++)
        {
            if (Ranked[i] == code)
                return i;
        }

        return null;
    }
}

File: MishkatulIlm-Client/src/app/core/utils/lesson-frequency-rules.ts
/** Ordered from lowest to highest lesson cadence (excludes FLEXIBLE). */
export const LESSON_FREQUENCY_RANKED = [
  'BIWEEKLY',
  'ONCE-WEEK',
  'TWICE-WEEK',
  'THREE-WEEK',
  'FOUR-PLUS-WEEK',
] as const;

export type RankedLessonFrequency = (typeof LESSON_FREQUENCY_RANKED)[number];

const ALL_FREQUENCY_CODES = [...LESSON_FREQUENCY_RANKED, 'FLEXIBLE'] as const;

/** Allowed frequency codes for the given subject count (excludes empty placeholder). */
export function allowedLessonFrequencyCodes(subjectCount: number): string[] {
  if (subjectCount === 3) {
    return ['TWICE-WEEK', 'THREE-WEEK'];
  }
  return [...ALL_FREQUENCY_CODES];
}

export function isLessonFrequencyAllowed(frequency: string, subjectCount: number): boolean {
  const code = frequency.trim().toUpperCase();
  if (!code) return false;
  return allowedLessonFrequencyCodes(subjectCount).includes(code);
}

export function lessonFrequencyConstraintHint(subjectCount: number): string | null {
  if (subjectCount === 3) {
    return 'With 3 subjects selected, choose twice or three times per week.';
  }
  return null;
}
